File: go-frontend/sim/euler.go
package sim

import(
  . "../tensor";
  . "math";
  //.  "fmt";
)

type Euler struct{
  m_tensor, h_tensor *Tensor4;
  m, h [][]float;	// components as lists
  field *FieldPlan;
  t, dt float;
}

func NewEuler(m *Tensor4, field *FieldPlan, dt float) *Euler{
  solver := new(Euler);
  
  solver.m_tensor = m;
  solver.h_tensor = NewTensor4(m.Size());
  // store m and h components as lists.
  solver.m, solver.h = make([][]float, 3), make([][]float, 3);
  for c:=0; c<3; c++{
      solver.m[c] = solver.m_tensor.Component(c).List();
      solver.h[c] = solver.h_tensor.Component(c).List();
  }

  solver.t = 0.;
  solver.dt = dt;
  solver.field = field;
  return solver;
}

func (s *Euler) Step(){
    alpha := 1.0;
    gilbert := 1.0 / (1.0 + alpha * alpha);
    
    s.field.Execute(s.m_tensor, s.h_tensor);
    m, h, torque := make([]float, 3), make([]float, 3), make([]float, 3);
    M := s.m;
    H := s.h;
    for i:=0; i<len(s.m[0]); i++{
      for c:=X; c<=Z; c++{
	m[c] = M[c][i];
	h[c] = H[c][i];
      }
      Torque(m, h, torque, alpha, gilbert);
      //Println(torque);
      norm:=0.;
      for c:=X; c<=Z; c++{
	m[c] += s.dt * torque[c];
	norm += m[c]*m[c];
      }
      norm = 1./float(Sqrt(float64(norm)));
      for c:=X; c<=Z; c++{
	M[c][i] = m[c] * norm;
      }
    }
}



func Torque(m, h, torque []float, alpha, gilbert float) {
    // - m cross H
     _mxHx := -m[Y] * h[Z] + h[Y] * m[Z];
     _mxHy := m[X] * h[Z] - h[X] * m[Z];
     _mxHz := -m[X] * h[Y] + h[X] * m[Y];

    // - m cross (m cross H)
     _mxmxHx := m[Y] * _mxHz - _mxHy * m[Z];
     _mxmxHy := -m[X] * _mxHz + _mxHx * m[Z];
     _mxmxHz := m[X] * _mxHy - _mxHx * m[Y];

    
    torque[X] = (_mxHx + _mxmxHx * alpha) * gilbert;
    torque[Y] = (_mxHy + _mxmxHy * alpha) * gilbert;
    torque[Z] = (_mxHz + _mxmxHz * alpha) * gilbert;
}

File: core/euler.cpp
#include "euler.h"

eulersolver* new_euler(tensor* m, convplan* field, double dt){
  eulersolver* s = (eulersolver*)malloc(sizeof(eulersolver)); 
  
  s->m_tensor = m;
  s->h_tensor = new_tensorN(4, m->size);
  
  s->m = (float**)calloc(3, sizeof(float*));
  s->h = (float**)calloc(3, sizeof(float*));
  for(int i=0; i<3; i++){
    s->m[i] = tensor_component(s->m_tensor, i)->list;
    s->h[i] = tensor_component(s->h_tensor, i)->list;
  }
  
  s->size = (int*)calloc(3, sizeof(int));
  s->size[0] = m->size[1];
  s->size[1] = m->size[2];
  s->size[2] = m->size[3];
  s->N = s->size[0] * s->size[1] * s->size[2];
  printf("new_euler: %d x %d x %d = %d\n", s->size[0], s->size[1], s->size[2], s->N);
  
  s->torque_buf = (float*)calloc(3, sizeof(float));
  
  s->t = 0.;
  s->dt = dt;
  s->field = field;
  
  return s;
}

void euler_step(eulersolver* s){
     float** m = s->m;
     float** h = s->h;

    double alpha = 1.0;
    double gilbert = 1.0 / (1.0 + alpha * alpha);
    
    conv_execute(s->field, s->m_tensor->list, s->h_tensor->list);
    
    printf("s->N=%d\n", s->N);
    for(int i=0; i < s->N; i++){
      torque(m[X][i], m[Y][i], m[Z][i], h[X][i], h[Y][i], h[Z][i], alpha, gilbert, s->torque_buf);
      printf("torque = %f, %f, %f\n", s->torque_buf[X], s->torque_buf[Y], s->torque_buf[Z]);
      float norm  = 0.;
      for(int c=X; c<=Z; c++){
	m[c][i] += s->dt * s->torque_buf[c];
	norm += m[c][i]*m[c][i];
      }
    
      norm = 1./(sqrt(norm));
      for(int c=X; c<=Z; c++){
	m[c][i] *= norm;
      }
      
    }
    
    s->t += s->dt;
}